azure_identity_access_token_provider.go

```diff
@@ -1,97 +1,132 @@
+func NewAzureIdentityAccessTokenProviderWithScopesAndValidHostsAndObservabilityOptions(credential azcore.TokenCredential, scopes []string, validHosts []string, observabilityOptions ObservabilityOptions) (*AzureIdentityAccessTokenProvider, error) {
+	if credential == nil {
+		return nil, errors.New("credential cannot be nil")
+	}
+	scopesLen := len(scopes)
+	finalScopes := make([]string, scopesLen)
+	if scopesLen == 0 {
+		finalScopes = append(finalScopes, "https://graph.microsoft.com/.default")
+	} else {
+		copy(finalScopes, scopes)
+	}
+	validator := absauth.NewAllowedHostsValidator(validHosts)
+	if len(validHosts) == 0 {
+		validator = absauth.NewAllowedHostsValidator([]string{"graph.microsoft.com", "graph.microsoft.us", "dod-graph.microsoft.us", "graph.microsoft.de", "microsoftgraph.chinacloudapi.cn", "canary.graph.microsoft.com"})
+	}
+	result := &AzureIdentityAccessTokenProvider{
+		credential:            credential,
+		scopes:                finalScopes,
+		allowedHostsValidator: &validator,
+		observabilityOptions:  observabilityOptions,
+	}
+
+	return result, nil
+}
+
+const claimsKey = "claims"
+
+// GetAuthorizationToken returns the access token for the provided url.
+func (p *AzureIdentityAccessTokenProvider) GetAuthorizationToken(ctx context.Context, url *u.URL, additionalAuthenticationContext map[string]interface{}) (string, error) {
+	ctx, span := otel.GetTracerProvider().Tracer(p.observabilityOptions.GetTracerInstrumentationName()).Start(ctx, "GetAuthorizationToken")
+	defer span.End()
+	if !(*(p.allowedHostsValidator)).IsUrlHostValid(url) {
+		span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.is_url_valid", false))
+		return "", nil
+	}
+	if !strings.EqualFold(url.Scheme, "https") {
+		span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.is_url_valid", false))
+		err := errors.New("url scheme must be https")
+		span.RecordError(err)
+		return "", err
+	}
+	span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.is_url_valid", true))
+
+	claims := ""
+
+	if additionalAuthenticationContext != nil &&
+		additionalAuthenticationContext[claimsKey] != nil {
+		if rawClaims, ok := additionalAuthenticationContext[claimsKey].(string); ok {
+			decodedClaims, err := base64.StdEncoding.DecodeString(rawClaims)
+			if err != nil {
+				span.RecordError(err)
+				return "", err
+			}
+			claims = string(decodedClaims)
+			err = errors.New("received a claim for CAE but azure identity doesn't support claims: " + claims + " https://github.com/Azure/azure-sdk-for-go/issues/14284")
+			span.RecordError(err)
+			return "", err
+		}
+	}
+	if claims == "" {
+		span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.additional_claims_provided", false))
+	} else {
+		span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.additional_claims_provided", true))
+	}

```

Comment:
nit:

```go
claimsProvided := true
if claims == "" {
      claimsProvided = false
}
span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.additional_claims_provided", claimsProvided))
```

Reply:
```suggestion
	span.SetAttributes(attribute.Bool("com.microsoft.kiota.authentication.additional_claims_provided", claims != ""))

```

better yet (no assignment)